// flowbite-react/lib/esm/components/Dropdown/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useMemo } from 'react';
import { HiOutlineChevronDown, HiOutlineChevronLeft, HiOutlineChevronRight, HiOutlineChevronUp } from 'react-icons/hi';
import { excludeClassName } from '../../helpers/exclude';
import { Button } from '../Button';
import { Tooltip } from '../Tooltip';
import { DropdownDivider } from './DropdownDivider';
import { DropdownHeader } from './DropdownHeader';
import { DropdownItem } from './DropdownItem';
const icons = {
    top: HiOutlineChevronUp,
    right: HiOutlineChevronRight,
    bottom: HiOutlineChevronDown,
    left: HiOutlineChevronLeft,
};
const DropdownComponent = (props) => {
    const theirProps = excludeClassName(props);
    const { children, label, inline, tooltipArrow = false, arrowIcon = true, ...restProps } = theirProps;
    const { placement = inline ? 'bottom-start' : 'bottom', trigger = 'click', ...buttonProps } = restProps;
    const Icon = useMemo(() => {
        const [p] = placement.split('-');
        return icons[p] ?? HiOutlineChevronDown;
    }, [placement]);
    const content = useMemo(() => _jsx("ul", { className: "py-1", children: children }), [children]);
    const TriggerWrapper = ({ children }) => inline ? _jsx("button", { className: "flex items-center", children: children }) : _jsx(Button, { ...buttonProps, children: children });
    return (_jsx(Tooltip, { content: content, style: "auto", animation: "duration-100", placement: placement, arrow: tooltipArrow, trigger: trigger, children: _jsxs(TriggerWrapper, { children: [label, arrowIcon && _jsx(Icon, { className: "ml-2 h-4 w-4" })] }) }));
};
DropdownComponent.displayName = 'Dropdown';
DropdownItem.displayName = 'Dropdown.Item';
DropdownHeader.displayName = 'Dropdown.Header';
DropdownDivider.displayName = 'Dropdown.Divider';
export const Dropdown = Object.assign(DropdownComponent, {
    Item: DropdownItem,
    Header: DropdownHeader,
    Divider: DropdownDivider,
});
