// flowbite-react/lib/cjs/components/Dropdown/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dropdown = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const hi_1 = require("react-icons/hi");
const exclude_1 = require("../../helpers/exclude");
const Button_1 = require("../Button");
const Tooltip_1 = require("../Tooltip");
const DropdownDivider_1 = require("./DropdownDivider");
const DropdownHeader_1 = require("./DropdownHeader");
const DropdownItem_1 = require("./DropdownItem");
const icons = {
    top: hi_1.HiOutlineChevronUp,
    right: hi_1.HiOutlineChevronRight,
    bottom: hi_1.HiOutlineChevronDown,
    left: hi_1.HiOutlineChevronLeft,
};
const DropdownComponent = (props) => {
    const theirProps = (0, exclude_1.excludeClassName)(props);
    const { children, label, inline, tooltipArrow = false, arrowIcon = true, ...restProps } = theirProps;
    const { placement = inline ? 'bottom-start' : 'bottom', trigger = 'click', ...buttonProps } = restProps;
    const Icon = (0, react_1.useMemo)(() => {
        const [p] = placement.split('-');
        return icons[p] ?? hi_1.HiOutlineChevronDown;
    }, [placement]);
    const content = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)("ul", { className: "py-1", children: children }), [children]);
    const TriggerWrapper = ({ children }) => inline ? (0, jsx_runtime_1.jsx)("button", { className: "flex items-center", children: children }) : (0, jsx_runtime_1.jsx)(Button_1.Button, { ...buttonProps, children: children });
    return ((0, jsx_runtime_1.jsx)(Tooltip_1.Tooltip, { content: content, style: "auto", animation: "duration-100", placement: placement, arrow: tooltipArrow, trigger: trigger, children: (0, jsx_runtime_1.jsxs)(TriggerWrapper, { children: [label, arrowIcon && (0, jsx_runtime_1.jsx)(Icon, { className: "ml-2 h-4 w-4" })] }) }));
};
DropdownComponent.displayName = 'Dropdown';
DropdownItem_1.DropdownItem.displayName = 'Dropdown.Item';
DropdownHeader_1.DropdownHeader.displayName = 'Dropdown.Header';
DropdownDivider_1.DropdownDivider.displayName = 'Dropdown.Divider';
exports.Dropdown = Object.assign(DropdownComponent, {
    Item: DropdownItem_1.DropdownItem,
    Header: DropdownHeader_1.DropdownHeader,
    Divider: DropdownDivider_1.DropdownDivider,
});
