// flowbite-react/lib/cjs/components/FormControls/Select.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Select = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classnames_1 = __importDefault(require("classnames"));
const colorClasses = {
    base: {
        input: 'bg-gray-50 border-gray-300 text-gray-900 focus:border-blue-500 focus:ring-blue-500 dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400 dark:focus:border-blue-500 dark:focus:ring-blue-500',
        helperText: 'text-gray-500 dark:text-gray-400',
    },
    green: {
        input: 'border-green-500 bg-green-50 text-green-900 placeholder-green-700 focus:border-green-500 focus:ring-green-500 dark:border-green-400 dark:bg-green-100 dark:focus:border-green-500 dark:focus:ring-green-500',
        helperText: 'text-green-600 dark:text-green-500',
    },
    red: {
        input: 'border-red-500 bg-red-50 text-red-900 placeholder-red-700 focus:border-red-500 focus:ring-red-500 dark:border-red-400 dark:bg-red-100 dark:focus:border-red-500 dark:focus:ring-red-500',
        helperText: 'text-red-600 dark:text-red-500',
    },
};
const Select = ({ children, className, sizing = 'md', shadow, helperText, addon, icon: Icon, color = 'base', ...props }) => ((0, jsx_runtime_1.jsxs)("div", { className: "flex", children: [addon && ((0, jsx_runtime_1.jsx)("span", { className: "inline-flex items-center rounded-l-md border border-r-0 border-gray-300 bg-gray-200 px-3 text-sm text-gray-900 dark:border-gray-600 dark:bg-gray-600 dark:text-gray-400", children: addon })), (0, jsx_runtime_1.jsxs)("div", { className: "relative w-full", children: [Icon && ((0, jsx_runtime_1.jsx)("div", { className: "pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3", children: (0, jsx_runtime_1.jsx)(Icon, { className: "h-5 w-5 text-gray-500 dark:text-gray-400" }) })), (0, jsx_runtime_1.jsx)("select", { className: (0, classnames_1.default)('block w-full border disabled:cursor-not-allowed disabled:opacity-50', colorClasses[color].input, {
                        'pl-10': Icon,
                        'rounded-lg': !addon,
                        'rounded-r-lg': addon,
                        'shadow-sm dark:shadow-sm-light': shadow,
                        'p-2 sm:text-xs': sizing === 'sm',
                        'p-2.5 text-sm': sizing === 'md',
                        'sm:text-md p-4': sizing === 'lg',
                    }, className), ...props, children: children }), helperText && (0, jsx_runtime_1.jsx)("p", { className: (0, classnames_1.default)('mt-2 text-sm', colorClasses[color].helperText), children: helperText })] })] }));
exports.Select = Select;
