// flowbite-react/lib/cjs/components/Footer/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Footer = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classnames_1 = __importDefault(require("classnames"));
const FooterBrand_1 = require("./FooterBrand");
const FooterCopyright_1 = require("./FooterCopyright");
const FooterIcon_1 = require("./FooterIcon");
const FooterLink_1 = require("./FooterLink");
const FooterLinkGroup_1 = require("./FooterLinkGroup");
const FooterComponent = ({ children, className }) => {
    return ((0, jsx_runtime_1.jsx)("footer", { "data-testid": "footer-element", className: (0, classnames_1.default)('w-full rounded-lg bg-white p-4 shadow dark:bg-gray-800 md:flex md:items-center md:justify-between md:p-6', className), children: children }));
};
FooterComponent.displayName = 'Footer';
FooterCopyright_1.FooterCopyright.displayName = 'Footer.Copyright';
FooterLink_1.FooterLink.displayName = 'Footer.Link';
FooterBrand_1.FooterBrand.displayName = 'Footer.Brand';
FooterLinkGroup_1.FooterLinkGroup.displayName = 'Footer.LinkGroup';
FooterIcon_1.FooterIcon.displayName = 'Footer.Icon';
exports.Footer = Object.assign(FooterComponent, {
    Copyright: FooterCopyright_1.FooterCopyright,
    Link: FooterLink_1.FooterLink,
    LinkGroup: FooterLinkGroup_1.FooterLinkGroup,
    Brand: FooterBrand_1.FooterBrand,
    Icon: FooterIcon_1.FooterIcon,
});
